[MSHARED-987] Make use of SISU

cstamas opened the pull request:
https://issues.apache.org/jira/browse/MSHARED-987

Drop all the reflection hoops and loops, and organize into api-provider layout. Should have NO IMPACT on consumers.

Implement using this SISU based pattern: https://github.com/cstamas/providers In short: SISU loads components lazily, and silently skips component that throws  CNFEx during load. Moreover, Selector "figures our" current environment, more narrowing used delegate.

Changes at high level:
* code reorganization: maven-artifact-transfer-api (new), maven-artifact-transfer-maven-3.0.x (new), maven-artifact-transfer-maven-3.1.x (new) and maven-artifact-transfer (old).
* each module defines it's own (provided) expected environment and cleanly code against it, no reflection, no black magic
* using SISU collection lookup and some minor logic (not yet fully done, works for 3.0/3.1 case but could be improved, 302 repo exception is NOT in yet).
* existing UTs unmodified pass
* existing ITs unmodified pass except for 3.0.5

Benefits:
* each provider module set up it's "own" classpath, no need for reflection and overlapping dependency mess
* dropping a version support is just dropping a module
* adding new version support is just adding a new module
* consumers of m-a-t needs no change

cstamas commented:
As mentioned above UTs pass OK
The ITs pass, **except for maven 3.0.5**, reason: it seems it was not yet honoring "sisu index" (`META-INF/sisu/javax.inject.Named`). (ProjectInstaller pass OK, as it does not invoke 3.0.x).

Now, **this is a reason** to drop 3.0.x line, as I don't want to write anymore `plexus.xml`, while we have SISU.

Essentially, this means that regarding "components" maven 3.0.x and 3.1.x line "speak different languages". Maven 3.1.x+ "knows" sisu-index, while maven 3.0.x does NOT. IMO, I see really no reason to support 3.0.x.

Please test this, inspect ITs and ping here.

cstamas commented:
but now to "drop 3.0.x", it is simple as dropping module for 3.0.x :wink:

michael-o commented on maven-3.0.x/pom.xml:
```diff
@@ -0,0 +1,148 @@
+
+  <properties>
+    <maven30x.version>3.0</maven30x.version>
+    <sonatypeAether.version>1.7</sonatypeAether.version>
```
Does not exist in Maven Central: https://search.maven.org/artifact/org.sonatype.aether/aether-api

cstamas replied:
? https://repo1.maven.org/maven2/org/sonatype/aether/aether-api/1.7/

cstamas replied:
updated, as maven went 3.0.5, sonatype aether MUST go in lockstep, is 1.13.1

michael-o commented on maven-3.0.x/pom.xml:
```diff
@@ -0,0 +1,148 @@
+  <name>Apache Maven Artifact Transfer Maven 3.0.x provider</name>
+
+  <properties>
+    <maven30x.version>3.0</maven30x.version>
```
We use Mavn 3.0.5 as bare minimum.

michael-o commented on maven-3.0.x/src/main/java/org/apache/maven/shared/transfer/repository/internal/Maven302RepositoryManager.java:
```diff
@@ -20,18 +20,24 @@
-class Maven302RepositoryManager
-    extends Maven30RepositoryManager
+@Singleton
+@Named( "maven-3.0.2" )
```
I think this can be dropped because we won't get lower than 3.0.5.

cstamas replied:
fixed, as maven 3.0.x went to maven 3.0.5 (but kept original code, just make other abstract and this one MAVEN_3_0_X)

michael-o commented on maven-3.1.x/pom.xml:
```diff
@@ -0,0 +1,148 @@
+  <name>Apache Maven Artifact Transfer Maven 3.1.x provider</name>
+
+  <properties>
+    <maven31x.version>3.1.0</maven31x.version>
```
I think this should be 3.1.1 too.

cstamas replied:
fixed

michael-o commented on maven-artifact-transfer-api/pom.xml:
```diff
@@ -0,0 +1,107 @@
+    <dependency>
+      <groupId>org.apache.maven</groupId>
+      <artifactId>maven-artifact</artifactId>
+      <version>3.0</version>
```
Move version to property?

cstamas replied:
moved

michael-o commented on maven-artifact-transfer-api/src/main/java/org/apache/maven/shared/transfer/artifact/deploy/internal/DefaultArtifactDeployer.java:
```diff
@@ -0,0 +1,80 @@
+    {
+        if ( request == null )
+        {
+            throw new IllegalArgumentException( "The parameter request is not allowed to be null." );
```
NPE

michael-o commented on maven-artifact-transfer-api/src/main/java/org/apache/maven/shared/transfer/artifact/deploy/internal/DefaultArtifactDeployer.java:
```diff
@@ -0,0 +1,80 @@
+        {
+            throw new IllegalArgumentException( "The parameter request is not allowed to be null." );
+        }
+        if ( mavenArtifacts == null )
```
NPE

michael-o commented on maven-artifact-transfer-api/src/main/java/org/apache/maven/shared/transfer/artifact/install/internal/DefaultArtifactInstaller.java:
```diff
@@ -0,0 +1,89 @@
+                         Collection<Artifact> mavenArtifacts ) throws ArtifactInstallerException
+    {
+        validateParameters( request, mavenArtifacts );
+        if ( localRepository == null )
```
NPE

michael-o commented on maven-artifact-transfer-api/src/main/java/org/apache/maven/shared/transfer/artifact/install/internal/DefaultArtifactInstaller.java:
```diff
@@ -0,0 +1,89 @@
+
+    private void validateParameters( ProjectBuildingRequest request, Collection<Artifact> mavenArtifacts )
+    {
+        if ( request == null )
```
NPE

michael-o commented on maven-artifact-transfer-api/src/main/java/org/apache/maven/shared/transfer/artifact/install/internal/DefaultArtifactInstaller.java:
```diff
@@ -0,0 +1,89 @@
+        {
+            throw new IllegalArgumentException( "The parameter request is not allowed to be null." );
+        }
+        if ( mavenArtifacts == null )
```
NPE

michael-o commented on maven-artifact-transfer-api/src/main/java/org/apache/maven/shared/transfer/artifact/resolve/internal/DefaultArtifactResolver.java:
```diff
@@ -0,0 +1,89 @@
+
+    private void validateParameters( ProjectBuildingRequest buildingRequest, Artifact mavenArtifact )
+    {
+        if ( buildingRequest == null )
```
NPE

michael-o commented on maven-artifact-transfer-api/src/main/java/org/apache/maven/shared/transfer/artifact/resolve/internal/DefaultArtifactResolver.java:
```diff
@@ -0,0 +1,89 @@
+        {
+            throw new IllegalArgumentException( "The parameter buildingRequest is not allowed to be null." );
+        }
+        if ( mavenArtifact == null )
```
NPE

michael-o commented on maven-artifact-transfer-api/src/main/java/org/apache/maven/shared/transfer/artifact/resolve/internal/DefaultArtifactResolver.java:
```diff
@@ -0,0 +1,89 @@
+
+    private void validateParameters( ProjectBuildingRequest buildingRequest, ArtifactCoordinate coordinate )
+    {
+        if ( buildingRequest == null )
```
NPE

michael-o commented on maven-artifact-transfer-api/src/main/java/org/apache/maven/shared/transfer/artifact/resolve/internal/DefaultArtifactResolver.java:
```diff
@@ -0,0 +1,89 @@
+        {
+            throw new IllegalArgumentException( "The parameter buildingRequest is not allowed to be null." );
+        }
+        if ( coordinate == null )
```
NPE

michael-o commented on maven-artifact-transfer-api/src/main/java/org/apache/maven/shared/transfer/dependencies/collect/internal/DefaultDependencyCollector.java:
```diff
@@ -0,0 +1,105 @@
+
+    private void validateBuildingRequest( ProjectBuildingRequest buildingRequest )
+    {
+        if ( buildingRequest == null )
```
NPE

michael-o commented on maven-artifact-transfer-api/src/main/java/org/apache/maven/shared/transfer/dependencies/collect/internal/DefaultDependencyCollector.java:
```diff
@@ -0,0 +1,105 @@
+
+    private void validateRoot( Object root )
+    {
+        if ( root == null )
```
NPE

michael-o commented on maven-artifact-transfer-api/src/main/java/org/apache/maven/shared/transfer/dependencies/resolve/internal/DefaultDependencyResolver.java:
```diff
@@ -0,0 +1,107 @@
+    private void validateParameters( ProjectBuildingRequest buildingRequest, DependableCoordinate coordinate )
+    {
+        validateBuildingRequest( buildingRequest );
+        if ( coordinate == null )
```
NPE

michael-o commented on maven-artifact-transfer-api/src/main/java/org/apache/maven/shared/transfer/dependencies/resolve/internal/DefaultDependencyResolver.java:
```diff
@@ -0,0 +1,107 @@
+    private void validateParameters( ProjectBuildingRequest buildingRequest, Model model )
+    {
+        validateBuildingRequest( buildingRequest );
+        if ( model == null )
```
NPE

michael-o commented on maven-artifact-transfer-api/src/main/java/org/apache/maven/shared/transfer/dependencies/resolve/internal/DefaultDependencyResolver.java:
```diff
@@ -0,0 +1,107 @@
+
+    private void validateBuildingRequest( ProjectBuildingRequest buildingRequest )
+    {
+        if ( buildingRequest == null )
```
NPE

michael-o commented on maven-artifact-transfer/pom.xml:
```diff
@@ -0,0 +1,188 @@
+  <name>Apache Maven Artifact Transfer</name>
+
+  <properties>
+    <downstreamMaven.version>3.1.0</downstreamMaven.version>
```
3.1.1

cstamas replied:
fixed

michael-o commented on maven-artifact-transfer/pom.xml:
```diff
@@ -0,0 +1,188 @@
+                <extraArtifact>org.apache.maven:apache-maven:3.2.5:tar.gz:bin</extraArtifact>
+                <extraArtifact>org.apache.maven:apache-maven:3.3.1:tar.gz:bin</extraArtifact>
+                <extraArtifact>org.apache.maven:apache-maven:3.3.9:tar.gz:bin</extraArtifact>
+                <extraArtifact>org.apache.maven:apache-maven:3.5.0:tar.gz:bin</extraArtifact>
```
3.5.4

cstamas replied:
fixed, and added 3.6.3 and 3.8.1 as well

michael-o reviewed (changes requested):
Massive, thank you very much!

cstamas commented:
General remark @michael-o : I did not made any other code change then:
* added ProjectBuildingRequest parameter to methods, removed some unneeded members (like remote repositories).
* removed reflection bits, replaced with direct calls
* removed some reflection utilities, replaced with direct calls

My point is, that those NPE comments of your are about _old code_, I did not introduce any logic/code change in that area, and IMO we can do those later as improvement, but for now I'd leave code as is (is really just "reshuffled" into modules + changes I wrote above).

cstamas commented:
Applied PR comments (sans NPE, see above why), all ITs pass except the 3.0.5  (as before, as it's sisu does not speak sisu index)

michael-o commented:
> 
> 
> General remark @michael-o : I did not made any other code change then:
> 
>     * added ProjectBuildingRequest parameter to methods, removed some unneeded members (like remote repositories).
> 
>     * removed reflection bits, replaced with direct calls
> 
>     * removed some reflection utilities, replaced with direct calls
> 
> 
> My point is, that those NPE comments of your are about _old code_, I did not introduce any logic/code change in that area, and IMO we can do those later as improvement, but for now I'd leave code as is (is really just "reshuffled" into modules + changes I wrote above).

I agree with that.

michael-o commented:
Can you explain why we need `Maven302RepositoryManager`  instead of just `Maven302RepositoryManager`?

cstamas commented:
> Can you explain why we need `Maven302RepositoryManager` instead of just `Maven302RepositoryManager`?

Again, old code that I did not touch (aside of making it component). I wanted to make this reshuffle least volatile re old code, keep it unchanged as much is posslble

cstamas commented:
So to say, "restructure" the code, make these modules and remove the reflection cruft (to make easier to maintain and more readable), and then we merge this PR, and then as 2nd iteration we go and solve those NPEs and other stuff we think are not needed (the two 3.0.x related repomanagers, are to be merged into one class actually). But again, _this PR is not that PR_ 😄

rfscholte commented:
I think I know what you've done. Whereas I started with the single jar approach,  you separated the code into modules.
I can't see the new structure, but I know there were sometimes even 3 implementation due to changes in Aether. How did you solve that?

cstamas commented:
> I think I know what you've done. Whereas I started with the single jar approach, you separated the code into modules.
> I can't see the new structure, but I know there were sometimes even 3 implementation due to changes in Aether. How did you solve that?

I found max two implementations (3.0 and 3.0.2) of RepositoryManager, the rest was "clean" (one 3.0 and one 3.1). So unsure where there are 3 implementations...

In short, all the DefaultXXX classes extends this one
https://github.com/apache/maven-artifact-transfer/blob/MSHARED-987-providers/maven-artifact-transfer-api/src/main/java/org/apache/maven/shared/transfer/support/DelegatorSupport.java

And Selector (currently very roughly implemented, but we can easily refine it once needed for more differentiation)
https://github.com/apache/maven-artifact-transfer/blob/MSHARED-987-providers/maven-artifact-transfer-api/src/main/java/org/apache/maven/shared/transfer/support/Selector.java

It uses two key things:
* SISU: on "collection lookup" (injecting map of components) will silently skip components that cannot be loaded due CNFEx (so org.sonatype using things will be silently omitted from delegate map in maven 3.1, and other way around, org.eclipse using things will be omitted in maven 3.0 (note: this does not stand for DIRECT lookup, ie. if you lookup single component by key, and you get CNFex, SISU will fail naturally).
* Selector: then "narrows" and selects "proper" component (by detected runtime) from the map

Note: sisu injection of map does NOT instantiate all found components, as map is "lazy" (sisu detail), so at the end, you end with ONLY ONE instantiated component doing the work for given runtime (components are created atomically when they are getted from map).

Re selection: it is simple (only 3.0 vs 3.1) as I found no case of finer selection needed (there was 3.0 vs 3.0.2, but we upped 3.0 to 3.0.5). But, if needed, we can create fine grained or even hierarchical versioning, and make Selector "smarter", but as I said, is currently not needed.

cstamas commented:
Added some cleanup (properties may need more) and for fun disabled 3.0.5 in ITs, and added 3.6.3 and 3.8.1, so ITs now run on "3.1.1", "3.2.5", "3.3.9", "3.5.4", "3.6.3", "3.8.1" and all pass OK

cstamas commented:
@rfscholte with latest changes all the ITs pass with 3.0.5 as well (so your request accomplished), so ITs now run on "3.0.5", "3.1.1", "3.2.5", "3.3.9", "3.5.4", "3.6.3", "3.8.1" and all pass OK. Please review.

cstamas commented:
Hint for plugin developers: IF your plugin is Maven 3.1+ AND you want to use artifact-transfer AND you care for planet, then just exclude bits and make some penguins float longer on icebergs:
```xml
    <dependency>
      <groupId>org.apache.maven.shared</groupId>
      <artifactId>maven-artifact-transfer</artifactId>
      <version>2.0.0-SNAPSHOT</version>
      <exclusions>
        <exclusion>
          <groupId>org.apache.maven.shared</groupId>
          <artifactId>maven-artifact-transfer-maven-3.0.x</artifactId>
        </exclusion>
      </exclusions>
    </dependency>
```
The rest of m-artifact-transfer should be unaffected.

cstamas commented:
Um, anyone?

michael-o commented:
I will check when I can...

michael-o commented:
I will check when I can...

slachiewicz commented:
tested with m-dependency-plugin, works OOB

cstamas commented:
@slachiewicz cool, thanks for testing, maybe drop here some more plugins that use m-a-t and have nice IT suite, so I could test some of those as well...

cstamas commented:
Note to myself: I'd do one more thing in this PR: move out things into own packages (the "provider" ones), as currently all I did is just move out things in their corresponding module retaining original packages. This means that the 3 (new) modules have overlapping packages that is not nice. Given maven-3.0 and maven-3.1 (and future ones) should contain only component implementations (of those interfaces in m-a-t-api), we could just put them all under `org.apache.maven.shared.transfer.internal.maven30` and `org.apache.maven.shared.transfer.internal.maven31` etc package.

WDYT?

cstamas commented:
Any other comment? I'd like to merge or drop this, is lingering for way too long...

rfscholte commented:
To me the change is too big to validate, so it must proof itself for the several plugins out there that depend on it.

cstamas commented:
> To me the change is too big to validate, so it must proof itself for the several plugins out there that depend on it.

Ok, @slachiewicz tested with m-dependency-p, which other plugins you have on your mind?

michael-o commented:
Maven Deploy Plugin must be tested.

michael-o commented:
Question to reviews: Do we really need Sisu support for Maven 3.0.x? I'd rather drop it and focus on 3.1.x+ only.

elilja commented:
> Question to reviews: Do we really need Sisu support for Maven 3.0.x? I'd rather drop it and focus on 3.1.x+ only.

Several other Maven components already set 3.1.1 as a minimum, so spending time here for 3.0.x would be waste of time, it seems

slachiewicz commented:
See comments here https://github.com/apache/maven-artifact-transfer/pull/23

cstamas commented:
Plugin full IT suites:
* m-dependency-p (just upped the versoin of m-a-t, as I guess @slachiewicz did): all ok
* m-install-p: all ok, needs changes https://github.com/apache/maven-install-plugin/pull/12
* m-deploy-p: fails same ITs that fail on master, needs changes https://github.com/apache/maven-deploy-plugin/pull/18

m-deploy-p failed ITs on master are these
[ERROR] The following builds failed:
[ERROR] *  deploy-attached-sources/pom.xml
[ERROR] *  MDEPLOY-212/pom.xml
[ERROR] *  MDEPLOY-213/pom.xml

cstamas commented:
So anything else needed here?

michael-o reviewed (approved):
There are so many changes that I will blindly trust you here.

cstamas commented:
> There are so many changes that I will blindly trust you here.

Agreed, that's why I want to move forward, as IMHO people will realize how _simpler_ this will be than the current master "swamp" of fragile reflection magic.

Thanks for approving. There are 2 approvals on this PR. Is it enough?

michael-o commented:
@cstamas Go ahead a merge it. I appreciate your high quality work.
